provider/server/exec.go

```diff
@@ -5,11 +5,150 @@ package server
 
 import (
 	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os/exec"
+	"strings"
+	"sync"
+	"time"
 
-	ori "github.com/ironcore-dev/ironcore/iri/apis/machine/v1alpha1"
+	"github.com/creack/pty"
+	"github.com/digitalocean/go-libvirt"
+	"github.com/go-logr/logr"
+	iri "github.com/ironcore-dev/ironcore/iri/apis/machine/v1alpha1"
+	remotecommandserver "github.com/ironcore-dev/ironcore/poollet/machinepoollet/iri/streaming/remotecommand"
+	libvirtutils "github.com/ironcore-dev/libvirt-provider/pkg/libvirt/utils"
+	"github.com/moby/term"
+	"k8s.io/client-go/tools/remotecommand"
 )
 
-func (s *Server) Exec(ctx context.Context, req *ori.ExecRequest) (*ori.ExecResponse, error) {
+const (
+	StreamCreationTimeout = 30 * time.Second
+	StreamIdleTimeout     = 2 * time.Minute
+)
+
+type executorExec struct {
+	Libvirt         *libvirt.Libvirt
+	ExecRequest     *iri.ExecRequest
+	VirshExecutable string
+}
+
+func (s *Server) Exec(ctx context.Context, req *iri.ExecRequest) (*iri.ExecResponse, error) {
+	machineID := req.MachineId
+	log := s.loggerFrom(ctx, "MachineID", machineID)
+
+	log.V(1).Info("Inserting request into cache")
+	token, err := s.execRequestCache.Insert(req)
+	if err != nil {
+		return nil, err
+	}
+
+	log.V(1).Info("Returning url with token")
+	return &iri.ExecResponse{
+		Url: s.buildURL("exec", token),
+	}, nil
+}
+
+func (s *Server) ServeExec(w http.ResponseWriter, req *http.Request, token string) {
+	ctx := req.Context()
+	log := logr.FromContextOrDiscard(ctx)
+
+	request, ok := s.execRequestCache.Consume(token)
+	if !ok {
+		log.V(1).Info("Rejecting unknown / expired token")
+		http.NotFound(w, req)
+		return
+	}
+
+	exec := executorExec{
+		Libvirt:         s.libvirt,
+		ExecRequest:     request,
+		VirshExecutable: s.virshExecutable,
+	}
+
+	handler, err := remotecommandserver.NewExecHandler(exec, remotecommandserver.ExecHandlerOptions{
+		StreamCreationTimeout: StreamCreationTimeout,
+		StreamIdleTimeout:     StreamIdleTimeout,
+	})
+	if err != nil {
+		log.Error(err, "Error creating exec handler")
+		code := http.StatusInternalServerError
+		http.Error(w, http.StatusText(code), code)
+		return
+	}
+
+	handler.Handle(w, req, remotecommandserver.ExecOptions{})
+}
+
+func (e executorExec) Exec(ctx context.Context, in io.Reader, out io.WriteCloser, resize remotecommand.TerminalSizeQueue) error {
+	var wg sync.WaitGroup
+
+	machineID := e.ExecRequest.MachineId
+	log := logr.FromContextOrDiscard(ctx).WithName(machineID)
+
+	_, err := e.Libvirt.DomainLookupByUUID(libvirtutils.UUIDStringToBytes(machineID))
+	if err != nil {
+		if !libvirtutils.IsErrorCode(err, libvirt.ErrNoDomain) {
+			return fmt.Errorf("error looking up domain: %w", err)
+		}
+		return fmt.Errorf("machine %s has not yet been synced", machineID)
+	}
```

Comment:
A more fine grained error handling would be nice. It might make sense to check in the store if: 
-  the machine is present
-  the machine is in an available state